geode-core/src/main/java/org/apache/geode/management/internal/cli/GfshParser.java

```diff
@@ -57,10 +57,14 @@ public GfshParser(CommandManager commandManager) {
       add(command);
     }
 
-    for (Converter<?> converter : commandManager.getConverters()) {
+    for (Converter converter : commandManager.getConverters()) {
       if (converter.getClass().isAssignableFrom(ArrayConverter.class)) {
         ArrayConverter arrayConverter = (ArrayConverter) converter;
-        arrayConverter.setConverters(new HashSet<>(commandManager.getConverters()));
+        HashSet<Converter<?>> converters = new HashSet<>();
+        for (Converter c : commandManager.getConverters()) {
+          converters.add(c);
+        }
+        arrayConverter.setConverters(converters);
```

Comment:
Surely the logic in these nested foreach's could be simplified.

Reply:
At the very least, that should be a `Set` and not a `HashSet` on the left.

If we wanted to collapse this into a `stream`, we could.

```
        Set<Converter<?>> converters = new HashSet<>();
        for (Converter c : commandManager.getConverters()) {
          converters.add(c);
        }
```
It seems to me that this is just doing some strange workaround to get a set of `Converter` to a set of `Converter<?>`.  That could be rewritten as
```
        Set<Converter<?>> converters = commandManager.getConverters().stream().map(c -> (Converter<?>) c).collect(
            Collectors.toSet());
```

But at that point, the whole loop would go from
```
    for (Converter converter : commandManager.getConverters()) {
      if (converter.getClass().isAssignableFrom(ArrayConverter.class)) {
        ArrayConverter arrayConverter = (ArrayConverter) converter;
        Set<Converter<?>> converters = commandManager.getConverters().stream().map(c -> (Converter<?>) c).collect(
            Collectors.toSet());
        arrayConverter.setConverters(converters);
      }
      add(converter);
    }
```
to
```
   commandManager.getConverters()
        .stream()
        .filter(converter -> converter.getClass().isAssignableFrom(ArrayConverter.class))
        .forEach(converter ->
            ((ArrayConverter) converter).setConverters(
              commandManager.getConverters().stream().map(
                  c -> (Converter<?>) c).collect(
                Collectors.toSet())));
    commandManager.getConverters().forEach(this::add);
```
with the caveat that the `add(converter)` is now happening for each after the others have been addressed, and not in the middle of the loop.  But I don't think that should really matter, should it..?

Looking at that, I don't really believe it to be more readable than the first code block in this comment.  So probably split the difference, with
```
    for (Converter converter : commandManager.getConverters()) {
      if (converter.getClass().isAssignableFrom(ArrayConverter.class)) {
        ArrayConverter arrayConverter = (ArrayConverter) converter;
        arrayConverter.setConverters(
            commandManager.getConverters().stream()
                .map(c -> (Converter<?>) c).collect(Collectors.toSet()));
      }
      add(converter);
    }
```

I don't know.  I guess you could one step farther and reduce to
```
    for (Converter converter : commandManager.getConverters()) {
      if (converter.getClass().isAssignableFrom(ArrayConverter.class)) {
        ((ArrayConverter) converter).setConverters(
            commandManager.getConverters().stream()
                .map(c -> (Converter<?>) c).collect(Collectors.toSet()));
      }
      add(converter);
    }
```
but I personally don't like one-lining casts and method calls.

...

This comment got a lot bigger than I originally intended...

Reply:
Also, just for the record: all of the above is predicated on a hazy understanding of the difference between `Converter` and `Converter<?>`.

Reply:
The need here appears to stem from the way generics are being used (possibly incorrectly). Ideally we should just be able to say `HashSet<Converter<?>> converters = new HashSet<>(commandManager.getConverters();`. However that doesn't work because `getConverters` returns a `Set<Converter>` and not a `Set<Converter<?>>`. Ultimately the `setConverter` call wants to take a `Set<Converter<?>>` as an argument. Hence the need for this inner loop to keep the compiler happy.